File: qt/multimedia/gen_qwavedecoder.cpp
// +build ignore

#include <QAudioFormat>
#include <QIODevice>
#include <QIODeviceBase>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWaveDecoder>
#include <qwavedecoder.h>
#include "gen_qwavedecoder.h"
class MiqtVirtualQWaveDecoder : public virtual QWaveDecoder {
public:
MiqtVirtualQWaveDecoder(QIODevice* device): QWaveDecoder(device) {};
MiqtVirtualQWaveDecoder(QIODevice* device, const QAudioFormat& format): QWaveDecoder(device, format) {};
MiqtVirtualQWaveDecoder(QIODevice* device, QObject* parent): QWaveDecoder(device, parent) {};
MiqtVirtualQWaveDecoder(QIODevice* device, const QAudioFormat& format, QObject* parent): QWaveDecoder(device, format, parent) {};

virtual ~MiqtVirtualQWaveDecoder() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QWaveDecoder::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QWaveDecoder_MetaObject(const_cast<MiqtVirtualQWaveDecoder*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QWaveDecoder::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QWaveDecoder::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QWaveDecoder_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QWaveDecoder::qt_metacast(param1);

	}
};
QWaveDecoder* QWaveDecoder_new(QIODevice* device) {
return new MiqtVirtualQWaveDecoder(device);
}
QWaveDecoder* QWaveDecoder_new2(QIODevice* device, QAudioFormat* format) {
return new MiqtVirtualQWaveDecoder(device, *format);
}
QWaveDecoder* QWaveDecoder_new3(QIODevice* device, QObject* parent) {
return new MiqtVirtualQWaveDecoder(device, parent);
}
QWaveDecoder* QWaveDecoder_new4(QIODevice* device, QAudioFormat* format, QObject* parent) {
return new MiqtVirtualQWaveDecoder(device, *format, parent);
}
void QWaveDecoder_virtbase(QWaveDecoder* src
, QIODevice** outptr_QIODevice
) {
*outptr_QIODevice = static_cast<QIODevice*>(src);
}
QMetaObject* QWaveDecoder_MetaObject(const QWaveDecoder* self) {
	return (QMetaObject*) self->metaObject();
}
void* QWaveDecoder_Metacast(QWaveDecoder* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QWaveDecoder_Tr(const char* s) {
	QString _ret = QWaveDecoder::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QAudioFormat* QWaveDecoder_AudioFormat(const QWaveDecoder* self) {
	return new QAudioFormat(self->audioFormat());
}
QIODevice* QWaveDecoder_GetDevice(QWaveDecoder* self) {
	return self->getDevice();
}
int QWaveDecoder_Duration(const QWaveDecoder* self) {
	return self->duration();
}
long long QWaveDecoder_HeaderLength() {
	qint64 _ret = QWaveDecoder::headerLength();
	return static_cast<long long>(_ret);
}
bool QWaveDecoder_Open(QWaveDecoder* self, int mode) {
	return self->open(static_cast<QIODevice::OpenMode>(mode));
}
void QWaveDecoder_Close(QWaveDecoder* self) {
	self->close();
}
bool QWaveDecoder_Seek(QWaveDecoder* self, long long pos) {
	return self->seek(static_cast<qint64>(pos));
}
long long QWaveDecoder_Pos(const QWaveDecoder* self) {
	qint64 _ret = self->pos();
	return static_cast<long long>(_ret);
}
long long QWaveDecoder_Size(const QWaveDecoder* self) {
	qint64 _ret = self->size();
	return static_cast<long long>(_ret);
}
bool QWaveDecoder_IsSequential(const QWaveDecoder* self) {
	return self->isSequential();
}
long long QWaveDecoder_BytesAvailable(const QWaveDecoder* self) {
	qint64 _ret = self->bytesAvailable();
	return static_cast<long long>(_ret);
}
void QWaveDecoder_FormatKnown(QWaveDecoder* self) {
	self->formatKnown();
}
void QWaveDecoder_connect_FormatKnown(QWaveDecoder* self, intptr_t slot) {
	MiqtVirtualQWaveDecoder::connect(self, static_cast<void (QWaveDecoder::*)()>(&QWaveDecoder::formatKnown), self, [=]() {
		miqt_exec_callback_QWaveDecoder_FormatKnown(slot);
	});
}
void QWaveDecoder_ParsingError(QWaveDecoder* self) {
	self->parsingError();
}
void QWaveDecoder_connect_ParsingError(QWaveDecoder* self, intptr_t slot) {
	MiqtVirtualQWaveDecoder::connect(self, static_cast<void (QWaveDecoder::*)()>(&QWaveDecoder::parsingError), self, [=]() {
		miqt_exec_callback_QWaveDecoder_ParsingError(slot);
	});
}
struct miqt_string QWaveDecoder_Tr2(const char* s, const char* c) {
	QString _ret = QWaveDecoder::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QWaveDecoder_Tr3(const char* s, const char* c, int n) {
	QString _ret = QWaveDecoder::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QWaveDecoder_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQWaveDecoder*>( (QWaveDecoder*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QWaveDecoder_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQWaveDecoder*)(self) )->virtualbase_MetaObject();
}
void QWaveDecoder_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQWaveDecoder*>( (QWaveDecoder*)(self) )->handle__Metacast = slot;
}
void* QWaveDecoder_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQWaveDecoder*)(self) )->virtualbase_Metacast(param1);
}
void QWaveDecoder_Delete(QWaveDecoder* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQWaveDecoder*>( self );
	} else {
		delete self;
	}
}

File: qt/multimedia/gen_qwavedecoder.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QWaveDecoder struct {
	h          uintptr
	isSubclass bool
}

// NewQWaveDecoder constructs a new QWaveDecoder object.
func NewQWaveDecoder(device *qt.QIODevice) *QWaveDecoder {
	ret := newQWaveDecoder(QWaveDecoder_new((*QIODevice)(device.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

// NewQWaveDecoder2 constructs a new QWaveDecoder object.
func NewQWaveDecoder2(device *qt.QIODevice, format *QAudioFormat) *QWaveDecoder {
	ret := newQWaveDecoder(QWaveDecoder_new2((*QIODevice)(device.UnsafePointer()), format.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQWaveDecoder3 constructs a new QWaveDecoder object.
func NewQWaveDecoder3(device *qt.QIODevice, parent *qt.QObject) *QWaveDecoder {
	ret := newQWaveDecoder(QWaveDecoder_new3((*QIODevice)(device.UnsafePointer()), (*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

// NewQWaveDecoder4 constructs a new QWaveDecoder object.
func NewQWaveDecoder4(device *qt.QIODevice, format *QAudioFormat, parent *qt.QObject) *QWaveDecoder {
	ret := newQWaveDecoder(QWaveDecoder_new4((*QIODevice)(device.UnsafePointer()), format.cPointer(), (*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

func (this *QWaveDecoder) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QWaveDecoder_MetaObject(this.h)))
}

func (this *QWaveDecoder) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QWaveDecoder_Metacast(this.h, param1_Cstring))
}

func QWaveDecoder_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QWaveDecoder_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QWaveDecoder) AudioFormat() *QAudioFormat {
	_goptr := newQAudioFormat(QWaveDecoder_AudioFormat(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QWaveDecoder) GetDevice() *qt.QIODevice {
	return qt.UnsafeNewQIODevice(unsafe.Pointer(QWaveDecoder_GetDevice(this.h)))
}

func (this *QWaveDecoder) Duration() int {
	return (int)(QWaveDecoder_Duration(this.h))
}

func QWaveDecoder_HeaderLength() int64 {
	return (int64)(QWaveDecoder_HeaderLength())
}

func (this *QWaveDecoder) Open(mode OpenModeFlag) bool {
	return (bool)(QWaveDecoder_Open(this.h, (int)(mode)))
}

func (this *QWaveDecoder) Close() {
	QWaveDecoder_Close(this.h)
}

func (this *QWaveDecoder) Seek(pos int64) bool {
	return (bool)(QWaveDecoder_Seek(this.h, (longlong)(pos)))
}

func (this *QWaveDecoder) Pos() int64 {
	return (int64)(QWaveDecoder_Pos(this.h))
}

func (this *QWaveDecoder) Size() int64 {
	return (int64)(QWaveDecoder_Size(this.h))
}

func (this *QWaveDecoder) IsSequential() bool {
	return (bool)(QWaveDecoder_IsSequential(this.h))
}

func (this *QWaveDecoder) BytesAvailable() int64 {
	return (int64)(QWaveDecoder_BytesAvailable(this.h))
}

func (this *QWaveDecoder) FormatKnown() {
	QWaveDecoder_FormatKnown(this.h)
}

func (this *QWaveDecoder) OnFormatKnown(slot func()) {
	QWaveDecoder_connect_FormatKnown(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QWaveDecoder_FormatKnown
func miqt_exec_callback_QWaveDecoder_FormatKnown(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QWaveDecoder) ParsingError() {
	QWaveDecoder_ParsingError(this.h)
}

func (this *QWaveDecoder) OnParsingError(slot func()) {
	QWaveDecoder_connect_ParsingError(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QWaveDecoder_ParsingError
func miqt_exec_callback_QWaveDecoder_ParsingError(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func QWaveDecoder_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QWaveDecoder_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QWaveDecoder_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QWaveDecoder_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QWaveDecoder) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QWaveDecoder_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QWaveDecoder) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QWaveDecoder_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QWaveDecoder_MetaObject
func miqt_exec_callback_QWaveDecoder_MetaObject(self QWaveDecoder, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QWaveDecoder{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QWaveDecoder) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QWaveDecoder_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QWaveDecoder) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QWaveDecoder_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QWaveDecoder_Metacast
func miqt_exec_callback_QWaveDecoder_Metacast(self QWaveDecoder, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QWaveDecoder{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
